python/demo/gmsh/demo_gmsh.py

```diff
@@ -5,33 +5,101 @@
 # SPDX-License-Identifier:    LGPL-3.0-or-later
 #
 # =====================================
-# Mesh generation using Gmsh and pygmsh
+# Mesh generation using the GMSH python API
 # =====================================
 
-import numpy as np
-import pygmsh
+from dolfinx.mesh import create_mesh, create_meshtags
+from dolfinx.io import XDMFFile, ufl_mesh_from_gmsh
+from dolfinx.cpp.io import perm_gmsh, extract_local_entities
+from dolfinx import cpp
 from mpi4py import MPI
+import gmsh
+import numpy as np
+
+
+def gather_gmsh_physical_entities():
```

Comment:
This function loops really complicated. Can it be simplified? 

At the lest, removing some line breaks and adding come comments would improve readability.

Reply:
We need this function to gather physical tags into meshes. For instance when using `BooleanFragements` with gmsh,  you get multiple volume entities that has to be loaded.
This splitting is very common with entities of lower dimensions, such as facets.
I can add in some comments to explain what is going on.

Reply:
Function has been slightly simplified and is now well documented. 
Should we make it part of the dolfinx.io?